### PR TITLE
fixes #9815 - Corrects group parameter cloning on hostgroup

### PR DESCRIPTION
**Pull request by cfouant**

### CODE REVIEW

**Comment by domcleal** [comment]
Please take a look at the test failures on the API controller.  A unit test to demonstrate the issue would also be useful.

**Comment by domcleal** [comment]
Is this something that deep_cloneable would handle automatically?  I'm comparing it to Host::Managed which simply does a `deep_clone(:include => [:host_parameters])` and doesn't need the accessor, to wipe reference_id etc.

Perhaps you could give it a quick try and just compare if that's a better solution.

**Comment by cfouant** [comment]
@domcleal - it worked nicely. I still have to skip the validation by passing flag, but the changes have been made. Thanks!

**Comment by domcleal** [comment]
That's an improvement, thanks.  Would it work to do the entire Hostgroup clone with deep_clone too, and not just the parameters?  It may even sort the cloned flag issue too, as it's not something we have to workaround with Host::Managed, which I find curious.

**Comment by cfouant** [comment]
@domcleal - Okay, was able to clone most of it using deep_clone, with the exception of config_groups. I kept the flag added for cloning to avoid hitting the validation error (although I noticed it does the same for Host::Managed - they save in the end, but feel like it shouldn't give an error outright - I can remove if you like). Lastly, there is a belongs_to_host association at the parameter level which had to be removed if you want to deep_clone the parameters. Everything else seemed to clone well. Let me know if you'd like further changes, thanks!

**Comment by domcleal** [comment]
That's much nicer.  I see what you mean about Host::Managed, that's interesting.  I think the cause is that validator, it's simply not correct to validate presence of an ID on the associated object at that stage, since it won't be known until it's saved.

Moving the generic validator into all of app/models/parameters/\* to check for the specific association (e.g. GroupParameter checks "hostgroup") works better, since it doesn't flag up an error with an unsaved hostgroup/host etc.  e.g. for GroupParameter:

```
validates :hostgroup, :presence => {:message => N_("parameters require an associated host group")}
```

And in Parameter, I removed both attr_accessors (and nested from app/views/common_parameters/_parameter.html.erb), the reference_id validator.  This also fixes the validation error shown on the host form without needing to set cloned=true there too.

Correct on removing belongs_to_host, since that's already in HostParameter, it shouldn't have been in Parameter too.

Not sure about config_groups, let's not unpick that one right now, the fix seems OK.

File: app/models/hostgroup.rb [hunk 1]
```diff
@@ -194,20 +194,11 @@ def root_pass
...
-
-    # Clone any parameters as well
-    self.group_parameters.each{|param| new.group_parameters << param.clone}
-    self.lookup_values.each do |lookup_value|
-      new_lookup_value = lookup_value.dup
-      new_lookup_value.match = "hostgroup=#{new.title}"
-      new.lookup_values << new_lookup_value
-    end
+    new.group_parameters.each {|param| param.cloned = true}
+    new.name = name
```
Comment (domcleal) [inline app/models/hostgroup.rb:200]: I'd say just remove this line, let it remain blank (due to :except above) like it does with hosts.

Comment (domcleal) [inline app/models/hostgroup.rb:200]: Oh, disregard, "name" is a local variable (method argument).

**Comment by domcleal** [comment]
> validates :hostgroup, :presence => {:message => N_("parameters require an associated host group")}

In fact, we can remove the custom message too in that case as it's more specific.  `validates :hostgroup, :presence => true` would be sufficient.

**Comment by cfouant** [comment]
@domcleal Thanks so much for your help. I've made the changes and done some preliminary testing. I added validations on locations and organization parameters as well, but I think that may have been unnecessary (not sure). I think moving the validation to the parameters was the right choice, I was thinking that myself, but honestly not sure how all the parts affect one another, so didn't want to jump the gun on that. This looks much cleaner though!

**Comment by domcleal** [comment]
Yeah, I can't say I was certain either, I think a lot of it was just old, bad quality code.  Removing the hacks and workarounds should make it better in future.

Looks like some of the unit tests need "nested" removing from them too, check the core Foreman failures.

**Comment by domcleal** [comment]
Thanks.  I noticed that Puppet class associations aren't cloning properly, and realised this is a problem with some of the UI helpers that was also causing host clones to break (http://projects.theforeman.org/issues/8521).  I've fixed that over in #2588 which should also help fix it for the host groups change in this PR.

**Comment by cfouant** [comment]
@domcleal it looks like #2588 has all my code, should we just close this one out?

**Comment by domcleal** [comment]
Yep, if you're happy.  Thanks!  I'll get back to the other shortly, I was a bit stuck on the return value of one of those methods...

**Comment by ares** [comment]
Replaced by https://github.com/theforeman/foreman/pull/2661

**Pull request closed without merging**